pkg/scheduler/DESIGN.md

```diff
@@ -0,0 +1,548 @@
+# Query Request Queue Design: Queue Splitting and Prioritization
+
+The `RequestQueue` subservice embedded into the scheduler process is responsible for
+all decisions regarding enqueuing and dequeuing of query requests.
+While the `RequestQueue`'s responsibilities are relatively broad, including management of
+querier-worker connection lifecycles and graceful startup/shutdown,
+the queuing logic is isolated into a "tree queue" structure and its associated queue algorithms.
+
+## Tree Queue: What and Why
+
+The "tree queue" structure serves the purpose of a discrete priority queue.
```

Comment:
I'm not married to the word "builds," but I think the phrase `serves the purpose of` can be replaced with just one word.
```suggestion
The "tree queue" structure builds a discrete priority queue.
```